Get sstables for compaction idempotency

bhalevy opened the pull request:
This series makes `time_window_compaction_strategy::get_sstables_for_compaction` idempotent by setting its `_last_expired_check` member only when it find no fully expired sstables.

With that, "compaction_manager: do not copy the table compaction_strategy object" fixes a potential inefficiency (see #11051)

Applies on top of e1e4a73793c50c9f9aa0ee0fcf0b484bcca24b2d

Fixes #11051

denesb commented on compaction/time_window_compaction_strategy.cc:
```diff
@@ -225,19 +225,20 @@ time_window_compaction_strategy::get_sstables_for_compaction(table_state& table_
     std::unordered_set<shared_sstable> expired;
```
s/branch new/branch new/

bhalevy replied:
Fixed in v2 (to "brand new")

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -353,7 +353,7 @@ class compaction_manager::major_compaction_task : public compaction_manager::tas
         // those are eligible for major compaction.
         auto* t = _compacting_table;
-        sstables::compaction_strategy cs = t->get_compaction_strategy();
+        auto& cs = t->get_compaction_strategy();
```
Maybe we should delete the copy constructor?

denesb replied:
(and copy assignment operator)

bhalevy replied:
Done in v2

raphaelsc replied:
I don't follow.

compaction_strategy is:

class compaction_strategy {
    ::shared_ptr<compaction_strategy_impl> _compaction_strategy_impl;

so copying strategy won't instantiate an actual copy of the state, but rather share the ownership of it.

bhalevy replied:
<blush>you're absolutely right</blush>
That said, I think that obfuscates the code that isn't WYSIWYG anymore.
I think it would be better to have a unique_ptr in class compaction_strategy and to keep an explicit shared / lw_shared pointer where needed so it would be clearer where we just get a shared pointer.

raphaelsc replied:
Yes, one of the problems with using shared ptr underneath is that you break the copy semantics. copy with shared ptr means that we're getting a sort of reference to the original state, not an actual copy of it.

denesb reviewed (approved):
Look good to me, but maybe a better way to make sure bugs like this don't happen in the future either is to disallow copying compaction strategy objects. Aren't they virtual anyway?
Since I don't know whether that is possible, I'll approve the PR as otherwise LGTM.

bhalevy commented:
In v2 (4c4a1e4e24):
- reordered patches
- Fixed typo in change log
- added: "abstract_replication_strategy: delete the copy constructor and assignment operator"

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1377/
